crates/chain/src/chain.rs

```
@@ -557,69 +598,125 @@ pub async fn start_irys_node(
+                    reth_shutdown_receiver
+                ),
             )).unwrap();
+
+            debug!("Waiting for the main actor thread to finish");
+            actor_main_thread_handle.join().unwrap();
+
+            task_manager.graceful_shutdown();
+
+            irys_storage::reth_provider::cleanup_provider(&irys_provider);
+
+            debug!("Reth thread finished");
         })?;
 
     // wait for the full handle to be send over by the actix thread
-    Ok(irys_node_handle_receiver.await?)
+    let mut node = irys_node_handle_receiver.await?;
+    node.reth_thread_handle = Some(reth_thread_handler.into());
+    Ok(node)
 }
 
 async fn start_reth_node<T: HasName + HasTableType>(
-    exec: TaskExecutor,
+    task_executor: TaskExecutor,
     chainspec: ChainSpec,
     irys_config: Arc<IrysNodeConfig>,
     tables: &[T],
     sender: oneshot::Sender<FullNode<RethNode, RethNodeAddOns>>,
     irys_provider: IrysRethProvider,
     latest_block: u64,
+    consensus_engine_shutdown_receiver: tokio::sync::mpsc::Receiver<()>,
+    main_actor_thread_shutdown_sender: tokio::sync::mpsc::Sender<()>,
+    consensus_engine_shutdown_sender: tokio::sync::mpsc::Sender<()>,
 ) -> eyre::Result<NodeExitReason> {
     let node_handle = irys_reth_node_bridge::run_node(
         Arc::new(chainspec),
-        exec,
+        task_executor,
         irys_config,
         tables,
         irys_provider,
         latest_block,
+        consensus_engine_shutdown_receiver,
     )
     .await?;
+    debug!("Reth node started");
     sender
         .send(node_handle.node.clone())
         .expect("unable to send reth node handle");
+
     let exit_reason = node_handle.node_exit_future.await?;
+
+    debug!("Sending shutdown signal to main actor thread");
+    main_actor_thread_shutdown_sender.try_send(()).unwrap();
+
+    debug!("Sending shutdown signal to consensus engine");
+    consensus_engine_shutdown_sender.try_send(()).unwrap();
```

rather ignore the error than unwrap, so you can continue processing of the further steps

---

Done 👍